cmd/virt-chroot/cgroup.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"runtime"
+	"syscall"
+
+	runc_fs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
+	runcFs2 "github.com/opencontainers/runc/libcontainer/cgroups/fs2"
+	runc_configs "github.com/opencontainers/runc/libcontainer/configs"
+
+	"kubevirt.io/kubevirt/pkg/virt-handler/cgroup"
+)
+
+func decodeResources(marshalledResourcesHash string) (*runc_configs.Resources, error) {
+	var unmarshalledResources runc_configs.Resources
+
+	marshalledResources, err := base64.StdEncoding.DecodeString(marshalledResourcesHash)
+	if err != err {
+		return nil, fmt.Errorf("cannot decode marshalled cgroups resources. "+
+			"encoded resources: %s. err: %v", marshalledResourcesHash, err)
+	}
+
+	err = json.Unmarshal(marshalledResources, &unmarshalledResources)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshall cgroups resources. "+
+			"marshalled resources: %s. err: %v", marshalledResources, err)
+	}
+
+	return &unmarshalledResources, err
+}
+
+func decodePaths(marshalledPathsHash string) (map[string]string, error) {
+	var unmarshalledPaths map[string]string
+
+	marshalledPaths, err := base64.StdEncoding.DecodeString(marshalledPathsHash)
+	if err != err {
+		return nil, fmt.Errorf("cannot decode marshalled cgroups paths. "+
+			"encoded paths: %s. err: %v", marshalledPathsHash, err)
+	}
+
+	err = json.Unmarshal(marshalledPaths, &unmarshalledPaths)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshall cgroups paths. "+
+			"marshalled paths: %s. err: %v", marshalledPaths, err)
+	}
+
+	return unmarshalledPaths, err
+}
+
+func setCgroupResources(paths map[string]string, resources *runc_configs.Resources, isRootless bool, isV2 bool) error {
+	config := &runc_configs.Cgroup{
+		Path:      cgroup.HostCgroupBasePath,
+		Resources: &runc_configs.Resources{},
+	}
+
+	var err error
+
+	if isV2 {
+		err = setCgroupResourcesV2(paths, resources, isRootless, config)
+	} else {
+		err = setCgroupResourcesV1(paths, resources, isRootless, config)
+	}
+
+	if err != nil {
+		return fmt.Errorf("cannot set cgroup resources. err: %v", err)
+	}
+
+	return nil
+}
+
+func setCgroupResourcesV1(paths map[string]string, resources *runc_configs.Resources, isRootless bool, config *runc_configs.Cgroup) error {
+	return RunWithChroot(cgroup.HostCgroupBasePath, func() error {
+		cgroupManager := runc_fs.NewManager(config, paths, isRootless)
+		return cgroupManager.Set(resources)
+	})
+}
+
+func setCgroupResourcesV2(paths map[string]string, resources *runc_configs.Resources, isRootless bool, config *runc_configs.Cgroup) error {
+	cgroupDirPath := paths[""]
+
+	cgroupManager, err := runcFs2.NewManager(config, cgroupDirPath, isRootless)
+	if err != nil {
+		return fmt.Errorf("cannot create cgroups v2 manager. err: %v", err)
+	}
+
+	err = cgroupManager.Set(resources)
```

Comment:
The usage of virt-chroot for defining cgroup rules in this PR was first used for v2 actually, since it has spc_t permissions (spc stand for Super Privileged Container. This is important here since we need SELinux permissions to create eBPF programs which are the building blocks of cgroups v2 subsystem).

In the recent commits I've extended virt-chroot to also serve cgroups v1 requests to overcome the chrooting problem we had (`syscal.Chroot` affects the whole process, so now it affects virt-chroot only, not all virt-handler's goroutines).

So you understand correctly that chrooting is required only for v1, but using `virt-chroot` is needed also for v2 :)

Reply:
Ah, I see. Thanks for clarification.